### ISSUE TITLE
Feature request: clear all variable tensors

### ISSUE THREAD

**Issue by Russell91** [tensorflow/tensorflow#859]
When using ipython notebooks, one frequently runs the same cell without restarting the kernel. If the cell contains code of the form:

    x = tf.get_variable('x', (1,))

the cell will run correctly the first time, but throw an error the second. Obviously this error is appropriate in python files, but really just gets in the way for notebooks. It would be great if something like tf.clear_all_variables() were implemented, such that I could have a cell with contents:

    tf.clear_all_variables()
    x = tf.get_variable('x', (1,))

that I could run over and over again. On a larger scale, tensorflow seems to be rather unfriendly to notebooks in general, frequently requiring kernel restarts for small changes. But this issue for definitely contributes to the greatest number of forced kernel restarts for me.

**Comment by lukaszkaiser** [comment]
I'm not sure about one thing: is your intention to clear the variables in the scope, but preserve them in the graph, or to clear the whole graph? In the first case, I think it's dangerous -- why call it "x" if you're re-doing x? I understand if you want to clear the whole graph, but that can be done with tf.graph().as_default(), no? Please, let me know more of the context so we can solve it!

**Comment by mrry** [comment]
[`tf.reset_default_graph()`](https://www.tensorflow.org/versions/master/api_docs/python/framework.html#reset_default_graph) is probably the closest thing we have to this functionality, although it blows away all ops and tensors, as well as all variables.

**Comment by Russell91** [comment]
I see, yea my intention is to indeed reset everything. It seems that these options will work. Thanks.

**Comment by 32nguyen** [comment]
Thank mrry, I was looking for this

**Comment by MartinThoma** [comment]
@mrry When I add `tf.reset_default_graph()` to [my code](https://github.com/MartinThoma/HASY/blob/master/scripts/keras_hasy_cv.py), I get

```
Traceback (most recent call last):
  File "keras_hasy_cv.py", line 78, in <module>
    verbose=1, validation_data=(test_x, test_y))
  File "/home/moose/.local/lib/python2.7/site-packages/keras/models.py", line 672, in fit
    initial_epoch=initial_epoch)
  File "/home/moose/.local/lib/python2.7/site-packages/keras/engine/training.py", line 1192, in fit
    initial_epoch=initial_epoch)
  File "/home/moose/.local/lib/python2.7/site-packages/keras/engine/training.py", line 892, in _fit_loop
    outs = f(ins_batch)
  File "/home/moose/.local/lib/python2.7/site-packages/keras/backend/tensorflow_backend.py", line 1898, in __call__
    session = get_session()
  File "/home/moose/.local/lib/python2.7/site-packages/keras/backend/tensorflow_backend.py", line 118, in get_session
    _initialize_variables()
  File "/home/moose/.local/lib/python2.7/site-packages/keras/backend/tensorflow_backend.py", line 270, in _initialize_variables
    sess.run(tf.variables_initializer(uninitialized_variables))
  File "/usr/local/lib/python2.7/dist-packages/tensorflow/python/client/session.py", line 766, in run
    run_metadata_ptr)
  File "/usr/local/lib/python2.7/dist-packages/tensorflow/python/client/session.py", line 951, in _run
    fetch_handler = _FetchHandler(self._graph, fetches, feed_dict_string)
  File "/usr/local/lib/python2.7/dist-packages/tensorflow/python/client/session.py", line 407, in __init__
    self._fetch_mapper = _FetchMapper.for_fetch(fetches)
  File "/usr/local/lib/python2.7/dist-packages/tensorflow/python/client/session.py", line 238, in for_fetch
    return _ElementFetchMapper(fetches, contraction_fn)
  File "/usr/local/lib/python2.7/dist-packages/tensorflow/python/client/session.py", line 274, in __init__
    'Tensor. (%s)' % (fetch, str(e)))
ValueError: Fetch argument <tensorflow.python.framework.ops.Operation object at 0x7f7fe0e2e550> cannot be interpreted as a Tensor. (Operation name: "init"
op: "NoOp"
input: "^convolution2d_3_W/Assign"
input: "^convolution2d_3_b/Assign"
input: "^prelu_3_alphas/Assign"
```

**Comment by yaroslavvb** [comment]
You have to recreate everything from scratch after resetting graph

**Comment by ihsanafredi** [comment]
Hi, I am facing something similar, I posted the problem on [stackoverflow](https://stackoverflow.com/questions/47657157/variables-of-tensorflow-generate-error-in-a-loop) as well. But no help. I am calling a function with CNN, so if i return from that function, it should delete all the variable and on return create new ones. But it does not. I tried to use answer by mrry that use tf.reset_default_graph(), but even that gives error that no graph found. Can any body guide how to solve this issue.
